sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala

```diff
@@ -82,86 +130,169 @@ case class SortMergeJoin(
          *
          * @return true if the search is successful, and false if the right iterator runs out of
          *         tuples.
          */
         private def nextMatchingPair(): Boolean = {
-          if (!stop && rightElement != null) {
-            // run both side to get the first match pair
-            while (!stop && leftElement != null && rightElement != null) {
-              val comparing = keyOrdering.compare(leftKey, rightKey)
+          if (!stop && streamedElement != null) {
+            // step 1: run both side to get the first match pair
+            while (!stop && streamedElement != null && bufferedElement != null) {
+              val comparing = keyOrdering.compare(streamedKey, bufferedKey)
               // for inner join, we need to filter those null keys
-              stop = comparing == 0 && !leftKey.anyNull
-              if (comparing > 0 || rightKey.anyNull) {
-                fetchRight()
-              } else if (comparing < 0 || leftKey.anyNull) {
-                fetchLeft()
+              stop = comparing == 0 && !streamedKey.anyNull
+              if (comparing > 0 || bufferedKey.anyNull) {
+                if (joinType == FullOuter) {
+                  // the join type is full outer and the buffered side has a row with no
+                  // join match, so we have a result row with streamed null with buffered
+                  // side as this row. Then we fetch next buffered element and go back.
+                  continueStreamed = false
+                  return true
+                } else {
+                  fetchBuffered()
+                }
+              } else if (comparing < 0 || streamedKey.anyNull) {
+                if (joinType == Inner) {
+                  fetchStreamed()
+                } else {
+                  // the join type is not inner and the streamed side has a row with no
+                  // join match, so we have a result row with this streamed row with buffered
+                  // null row. Then we fetch next streamed element and go back.
+                  continueStreamed = true
+                  return true
+                }
               }
             }
-            rightMatches = new CompactBuffer[InternalRow]()
+            // step 2: run down the buffered side to put all matched rows in a buffer
+            bufferedMatches = new CompactBuffer[InternalRow]()
+            secondBufferedMatches = new CompactBuffer[InternalRow]()
             if (stop) {
               stop = false
               // iterate the right side to buffer all rows that matches
               // as the records should be ordered, exit when we meet the first that not match
-              while (!stop && rightElement != null) {
-                rightMatches += rightElement
-                fetchRight()
-                stop = keyOrdering.compare(leftKey, rightKey) != 0
+              while (!stop) {
+                if (boundCondition(joinRow(streamedElement, bufferedElement))) {
```

Comment:
May be we also need to keep the row when key match but `boundCondition ` return false for inner join.The example like above 
```sql
select a.key,b.key,a.value-b.value from a join b on a.key=b.key and a.value - b.value > 1
```